src/group_theory/monoid_localization.lean

```diff
@@ -6,410 +6,294 @@ Authors: Amelia Livingston
-
-end submonoid
 
 namespace add_submonoid
+variables {M : Type*} [add_comm_monoid M] (S : add_submonoid M) (N : Type*) [add_comm_monoid N]
 
-variables [add_comm_monoid X]
-
-/-- An alternate form of the `add_monoid` localization relation, stated here for readability of the
-    next few lemmas. -/
-private def r'.rel (Y : add_submonoid X) (a b : X × Y) :=
-∃ c : Y, (c : X) + (a.1 + b.2) = c + (b.1 + a.2)
-
-lemma r'.transitive {Y : add_submonoid X} : transitive (r'.rel Y) :=
-λ a b c ⟨m, hm⟩ ⟨n, hn⟩, ⟨n + m + b.2,
-  calc
-    ↑n + ↑m + ↑b.2 + (a.1 + ↑c.2)
-      = ↑n + (↑m + (b.1 + ↑a.2)) + ↑c.2 : by rw ←hm; abel
-  ... = ↑m + (↑n + (c.1 + ↑b.2)) + ↑a.2 : by rw ←hn; abel
-  ... = ↑n + ↑m + ↑b.2 + (c.1 + ↑a.2) : by abel⟩
-
-lemma r'.add {Y : add_submonoid X} {a b c d} :
-  r'.rel Y a b → r'.rel Y c d → r'.rel Y (a + c) (b + d) :=
-λ ⟨m, hm⟩ ⟨n, hn⟩, ⟨m + n,
-  calc
-    ↑m + ↑n + (a.1 + c.1 + (↑b.2 + ↑d.2))
-      = ↑n + c.1 + (↑m + (b.1 + ↑a.2)) + ↑d.2 : by rw ←hm; abel
-  ... = (↑m + (b.1 + ↑a.2)) + (↑n + (d.1 + ↑c.2)) : by rw ←hn; abel
-  ... = ↑m + ↑n + (b.1 + d.1 + (↑a.2 + ↑c.2)) : by abel⟩
-
-/-- An alternate form of the congruence relation on `X × Y`, `X` an `add_comm_monoid` and `Y` an
-    `add_submonoid` of `X`, whose quotient is the localization of `X` at `Y`. Its equivalence to
-    `r` can be useful for proofs. -/
-def r' (Y : add_submonoid X) : add_con (X × Y) :=
-{ r := λ a b, ∃ c : Y, (c : X) + (a.1 + b.2) = c + (b.1 + a.2),
-  iseqv := ⟨λ _, ⟨0, rfl⟩, λ _ _ ⟨c, hc⟩, ⟨c, hc.symm⟩, r'.transitive⟩,
-  add' := λ a b c d, r'.add }
+/-- The type of add_monoid homomorphisms satisfying the characteristic predicate: if `f : M →+ N`
+    satisfies this predicate, then `N` is isomorphic to the localization of `M` at `S`. -/
+structure localization_map :=
+(to_fun : M →+ N)
+(map_add_units : ∀ y : S, is_add_unit (to_fun y))
+(surj : ∀ z : N, ∃ x : M × S, z + to_fun x.2 = to_fun x.1)
+(eq_iff_exists : ∀ x y, to_fun x = to_fun y ↔ ∃ c : S, x + c = y + c)
 
 end add_submonoid
 
-variables [comm_monoid X] (Y : submonoid X) {Z : Type*} [comm_monoid Z]
+variables {M : Type*} [comm_monoid M] (S : submonoid M) (N : Type*) [comm_monoid N]
+          {P : Type*} [comm_monoid P]
 
 namespace submonoid
 
-/-- An alternate form of the congruence relation on `X × Y`, `X` a `comm_monoid` and `Y` a
-    submonoid of `X`, whose quotient is the localization of `X` at `Y`. Its equivalence to `r` can
+/-- The type of monoid homomorphisms satisfying the characteristic predicate: if `f : M →* N`
+    satisfies this predicate, then `N` is isomorphic to the localization of `M` at `S`. -/
+structure localization_map :=
+(to_fun : M →* N)
+(map_units : ∀ y : S, is_unit (to_fun y))
+(surj : ∀ z : N, ∃ x : M × S, z * to_fun x.2 = to_fun x.1)
+(eq_iff_exists : ∀ x y, to_fun x = to_fun y ↔ ∃ c : S, x * c = y * c)
+
+attribute [to_additive add_submonoid.localization_map] submonoid.localization_map
+
+@[to_additive] instance localization_map.inhabited :
```

Comment:
What is the reason for this instance?

Reply:
Probably because the linter complained.  I should probably add a comment to the linter output, but this is how we've been handling it sofar:
 * Structures like `nnreal`, `prod`, `list`, etc. should get `inhabited` instances so that you can use them with `list.inth` and other functions that require inhabitedness.
 * Types of morphism typically also get an inhabitedness instance (e.g. the constant zero function is typically a good choice)
 * Types of equivalences (and other Type-valued "propositions") get a `@[nolint has_inhabited_instance]`.

In all cases is the inhabited instance supposed to be as general as possible.

Reply:
So in this case the identity map from `R` to `R`? Or should we rather use the `nolint` attribute?

Reply:
If there is a canonical inhabitant (for any `M` and `N`) then that should be the inhabited instance.  The identity map is only a localization map if `M = N` is a group, right?  So I think we should just put a nolint there.

Reply:
The identity map is a localization map for the trivial submonoid, so I've used that for the instance.